src/render/shaders/Textures.hpp

```diff
@@ -213,35 +313,87 @@ void main() {
 }
 )#";
 
+inline const std::string FRAGBLURPREPARE = R"#(
+precision         mediump float;
+varying vec2      v_texcoord; // is in 0-1
+uniform sampler2D tex;
+
+uniform float     contrast;
+uniform float     brightness;
+
+void main() {
+    vec4  pixColor = texture2D(tex, v_texcoord);
+
+    // contrast
+    pixColor.rgb = (pixColor.rgb - 0.5) * contrast + 0.5;
+    // brightness
+    pixColor.rgb *= brightness;
+
+    gl_FragColor = pixColor;
+}
+)#";
+
 inline const std::string FRAGBLURFINISH = R"#(
-precision mediump float;
-varying vec2 v_texcoord; // is in 0-1
+precision         mediump float;
+varying vec2      v_texcoord; // is in 0-1
 uniform sampler2D tex;
 
-uniform float contrast;
-uniform float noise;
-uniform float brightness;
+uniform float     noise;
+uniform float     min_brightness;
+uniform float     max_brightness;
 
 float hash(vec2 p) {
-	return fract(sin(dot(p, vec2(12.9898, 78.233))) * 43758.5453);
+    return fract(sin(dot(p, vec2(12.9898, 78.233))) * 43758.5453);
+}
+
+vec3 rgb2hsv(vec3 c) {
+    vec4  K = vec4(0.0, -1.0 / 3.0, 2.0 / 3.0, -1.0);
+    vec4  p = mix(vec4(c.bg, K.wz), vec4(c.gb, K.xy), step(c.b, c.g));
+    vec4  q = mix(vec4(p.xyw, c.r), vec4(c.r, p.yzx), step(p.x, c.r));
+
+    float d = q.x - min(q.w, q.y);
+    float e = 1.0e-10;
+    return vec3(abs(q.z + (q.w - q.y) / (6.0 * d + e)), d / (q.x + e), q.x);
+}
+
+vec3 hsv2rgb(vec3 c) {
+    vec4 K = vec4(1.0, 2.0 / 3.0, 1.0 / 3.0, 3.0);
+    vec3 p = abs(fract(c.xxx + K.xyz) * 6.0 - K.www);
+    return c.z * mix(K.xxx, clamp(p - K.xxx, 0.0, 1.0), c.y);
+}
+
+float smoothClamp(float x, float a, float b) {
+    return smoothstep(0., 1., (x - a)/(b - a))*(b - a) + a;
 }
 
+float softClamp(float x, float min, float max) {
+    // Prevent setting 0 in the config to affect the brightness
+    // But also prevent weird banding  here
+    if (( min == 0. && max > 0.5) && x <= 0.45 ) return x;
+    if (( max == 0. && min < 0.5 ) && x >= 0.55 ) return x;
+
+    return smoothstep(0., 1., (2./3.)*(x - min)/(max - min) + (1./6.))*(max - min) + min;
+}
+
+
 void main() {
-    vec4 pixColor = texture2D(tex, v_texcoord);
+    vec4  pixColor = texture2D(tex, v_texcoord);
 
-    // contrast
-    pixColor.rgb = (pixColor.rgb - 0.5) * contrast + 0.5;
+    vec3  hsv = rgb2hsv(pixColor.rgb);
```

Comment:
these need a nope if no clamping is to be done

Reply:
Added in [b8fe17f](https://github.com/hyprwm/Hyprland/pull/3650/commits/b8fe17f0eb03f1763be1dceae3e21ab6537ca32a)